common/lib/xmodule/xmodule/vertical_block.py

```diff
@@ -177,3 +219,32 @@ def index_dictionary(self):
         xblock_body["content_type"] = "Sequence"
 
         return xblock_body
+
+    def find_descendent(self, block_key):
+        """
+        Return the descendent block with the given block key if it exists.
+
+        Otherwise return None.
+        """
+        for block in self.get_display_items():
+            if block.location == block_key:
+                return block
+
+    @XBlock.json_handler
+    def publish_event(self, data, suffix=''):  # pylint: disable=unused-argument
+        """
+        Publish data from the front end.
+        """
+        event_type = data.pop('event_type')
+        if event_type == 'grade':
+            # This handler can be called from the browser. Don't allow the browser to submit arbitrary grades ;-)
+            raise JsonHandlerError(400, "Posting grade events from the browser is forbidden.")
+        if event_type == 'completion':
+            block_key = UsageKey.from_string(data.pop('block_key'))
+            block = self.find_descendent(block_key)
+            if block is None:
+                raise JsonHandlerError(400, "Invalid block: %s" % block_key)
```

Comment:
Should we also check that this block is in the set returned by `get_completables_for_user` ? Or do we not care if spurious completion events or premature ones are emitted (e.g. user calling this directly to mark a gradable item as completed, instead of completing the graded activity).

Reply:
I think it doesn't matter if it's in the set of get_completables_for_user.  The main purpose of the specific user check is to reduce network traffic, and at this point, the request is already in the VPC (and will not even hit the DB if it doesn't change the value).

Reply:
Hmm.  But it might be worth the looser check of whether it uses the default completion method.  I'll think about that.